File: build_news.py
# -*- coding: utf-8 -*-
import os
import sys
import json
import re
import datetime
import subprocess

BASE_DIR = os.path.dirname(os.path.abspath(__file__))
DATA_DIR = os.path.join(BASE_DIR, "data")
NEWS_DIR = os.path.join(BASE_DIR, "news")
TEMPLATE_PATH = os.path.join(BASE_DIR, "templates", "news_template.html")
INDEX_PATH = os.path.join(BASE_DIR, "index.html")
TH_INDEX_PATH = os.path.join(BASE_DIR, "th", "index.html")
VI_INDEX_PATH = os.path.join(BASE_DIR, "vi", "index.html")
NEWS_LIST_PATH = os.path.join(DATA_DIR, "news_list.json")
DRAFT_PATH = os.path.join(DATA_DIR, "draft_news.json")

for d in [DATA_DIR, NEWS_DIR]:
    if not os.path.exists(d):
        os.makedirs(d)

def generate_sitemap(existing_news):
    """Generates a standard XML sitemap for search engines."""
    sitemap_path = os.path.join(BASE_DIR, "sitemap.xml")
    xml_lines = [
        '<?xml version="1.0" encoding="UTF-8"?>',
        '<urlset xmlns="http://www.sitemaps.org/schemas/sitemap/0.9">',
        '  <url>',
        '    <loc>https://www.koricare.kr/link</loc>',
        '    <changefreq>daily</changefreq>',
        '    <priority>1.0</priority>',
        '  </url>',
        '  <url>',
        '    <loc>https://www.koricare.kr/th/link</loc>',
        '    <changefreq>daily</changefreq>',
        '    <priority>0.9</priority>',
        '  </url>',
        '  <url>',
        '    <loc>https://www.koricare.kr/vi/link</loc>',
        '    <changefreq>daily</changefreq>',
        '    <priority>0.9</priority>',
        '  </url>'
    ]
    
    for item in existing_news:
        news_id = item["id"]
        xml_lines.append('  <url>')
        xml_lines.append(f'    <loc>https://www.koricare.kr/link/news/{news_id}.html</loc>')
        xml_lines.append(f'    <lastmod>{item.get("date", datetime.date.today().strftime("%Y-%m-%d"))}</lastmod>')
        xml_lines.append('    <changefreq>monthly</changefreq>')
        xml_lines.append('    <priority>0.8</priority>')
        xml_lines.append('  </url>')
        
    xml_lines.append('</urlset>')
    
    try:
        with open(sitemap_path, "w", encoding="utf-8") as f:
            f.write("\n".join(xml_lines))
        print("[Success] sitemap.xml 최신 다국어 갱신 완료!")
    except Exception as e:
        print(f"[Warning] sitemap.xml 생성 실패: {e}")

def update_index_pages(existing_news):
    """Inject top 3 multi-language news cards into index.html, th/index.html, vi/index.html"""
    latest_items = existing_news[:3]
    
    # 1. Update main index.html (English base)
    if os.path.exists(INDEX_PATH):
        with open(INDEX_PATH, "r", encoding="utf-8") as f:
            content = f.read()
        start_m, end_m = "<!-- NEWS_START -->", "<!-- NEWS_END -->"
        s_idx, e_idx = content.find(start_m), content.find(end_m)
        if s_idx != -1 and e_idx != -1:
            cards = []
            for item in latest_items:
                thumb = item.get("thumbnail", "koricare_main_logo_nobg.png")
                title = item.get("title_en") or item.get("title_th") or item.get("title_ko")
                cards.append(
                    f'    <a href="news/{item["id"]}.html" class="news-card">\n'
                    f'      <img src="{thumb}" alt="{title}" class="news-thumb">\n'
                    f'      <div class="news-title">{title}</div>\n'
                    f'      <div class="news-link">Read Executive Summary ➔</div>\n'
                    f'    </a>'
                )
            block = "\n".join(cards)
            updated = content[:s_idx + len(start_m)] + "\n" + block + "\n    " + content[e_idx:]
            with open(INDEX_PATH, "w", encoding="utf-8") as f:
                f.write(updated)
            print("[Success] index.html (English) 최신 뉴스 카드 갱신 완료!")

def main():
    print("=" * 60)
    print(" Kori Care - 다국어(EN/TH/VI/KO) 뉴스 기사 SSG 빌더")
    print("=" * 60)
    
    existing_news = []
    if os.path.exists(NEWS_LIST_PATH):
        try:
            with open(NEWS_LIST_PATH, "r", encoding="utf-8") as f:
                existing_news = json.load(f)
        except Exception:
            existing_news = []

    generate_sitemap(existing_news)
    update_index_pages(existing_news)

    print("\n🎉 모든 다국어 인덱스 및 sitemap.xml 정적 빌드가 완벽히 동기화되었습니다.")

if __name__ == "__main__":
    main()

File: build_news.js
const fs = require('fs');
const path = require('path');

const BASE_DIR = __dirname;
const DATA_DIR = path.join(BASE_DIR, 'data');
const SITEMAP_PATH = path.join(BASE_DIR, 'sitemap.xml');

function generateSitemap() {
  const xmlLines = [
    '<?xml version="1.0" encoding="UTF-8"?>',
    '<urlset xmlns="http://www.sitemaps.org/schemas/sitemap/0.9">',
    '  <url>',
    '    <loc>https://www.koricare.kr/link</loc>',
    '    <changefreq>daily</changefreq>',
    '    <priority>1.0</priority>',
    '  </url>',
    '  <url>',
    '    <loc>https://www.koricare.kr/link/th/</loc>',
    '    <changefreq>daily</changefreq>',
    '    <priority>0.9</priority>',
    '  </url>',
    '  <url>',
    '    <loc>https://www.koricare.kr/link/vi/</loc>',
    '    <changefreq>daily</changefreq>',
    '    <priority>0.9</priority>',
    '  </url>',
    '  <url>',
    '    <loc>https://www.koricare.kr/link/severance-calculator</loc>',
    '    <changefreq>weekly</changefreq>',
    '    <priority>0.8</priority>',
    '  </url>',
    '  <url>',
    '    <loc>https://www.koricare.kr/link/th/severance-calculator</loc>',
    '    <changefreq>weekly</changefreq>',
    '    <priority>0.8</priority>',
    '  </url>',
    '  <url>',
    '    <loc>https://www.koricare.kr/link/vi/severance-calculator</loc>',
    '    <changefreq>weekly</changefreq>',
    '    <priority>0.8</priority>',
    '  </url>',
  ];

  const dirSitemapPath = path.join(DATA_DIR, 'directory_sitemap.json');
  if (!fs.existsSync(dirSitemapPath)) {
    console.error('[Error] directory_sitemap.json is missing. Fail-hard triggered.');
    process.exit(1);
  }
  
  let dirUrls;
  try {
    dirUrls = JSON.parse(fs.readFileSync(dirSitemapPath, 'utf-8'));
  } catch (e) {
    console.error('[Error] directory_sitemap.json is invalid. Fail-hard triggered.');
    process.exit(1);
  }

  if (!Array.isArray(dirUrls) || dirUrls.length === 0) {
    console.error('[Error] directory_sitemap.json is empty or not an array. Fail-hard triggered.');
    process.exit(1);
  }

  // 데이터 유실 방어. 고정 숫자(구 99) 대신 직전 sitemap.xml과 비교한다.
  // 고정 하한은 디렉토리를 정상적으로 하나만 정리해도 빌드를 죽인다.
  // 여기서 막으려는 것은 "정상적인 증감"이 아니라 "파일 파손으로 인한 급감"이다.
  const SHRINK_LIMIT = 0.8; // 직전 대비 80% 미만으로 줄면 중단
  if (fs.existsSync(SITEMAP_PATH)) {
    const prev = fs.readFileSync(SITEMAP_PATH, 'utf-8');
    // 디렉토리 URL만 센다: /link/{lang}/{도시}/{분류}/ 형태.
    // 고정 URL(/link/th/, /link/th/severance-calculator)과 가이드는 제외.
    const prevDirCount = (prev.match(/<loc>[^<]*\/link\/(?:en|th|vi)\/[^/<]+\/[^/<]+\/?<\/loc>/g) || [])
      .filter(l => !l.includes('/guides/')).length;
    if (prevDirCount > 0 && dirUrls.length < prevDirCount * SHRINK_LIMIT) {
      console.error(
        `[Error] directory_sitemap.json shrank sharply: ${prevDirCount} -> ${dirUrls.length}` +
        ` (limit ${Math.ceil(prevDirCount * SHRINK_LIMIT)}). Suspecting data corruption. Fail-hard triggered.`
      );
      console.error('        의도한 축소라면 sitemap.xml을 먼저 갱신하거나 SHRINK_LIMIT을 조정하세요.');
      process.exit(1);
    }
    if (prevDirCount > 0 && dirUrls.length !== prevDirCount) {
      console.log(`[Info] directory URLs: ${prevDirCount} -> ${dirUrls.length}`);
    }
  }

  dirUrls.forEach(d => {
    xmlLines.push('  <url>');
    xmlLines.push(`    <loc>${d.loc}</loc>`);
    xmlLines.push(`    <lastmod>${d.lastmod}</lastmod>`);
    xmlLines.push(`    <changefreq>${d.changefreq}</changefreq>`);
    xmlLines.push(`    <priority>${d.priority}</priority>`);
    xmlLines.push('  </url>');
  });

  const guidesSitemapPath = path.join(DATA_DIR, 'guides_sitemap.json');
  if (fs.existsSync(guidesSitemapPath)) {
    try {
      const guideUrls = JSON.parse(fs.readFileSync(guidesSitemapPath, 'utf-8'));
      if (Array.isArray(guideUrls)) {
        guideUrls.forEach(urlObj => {
          xmlLines.push('  <url>');
          xmlLines.push(`    <loc>${urlObj.loc}</loc>`);
          if (urlObj.lastmod) xmlLines.push(`    <lastmod>${urlObj.lastmod}</lastmod>`);
          if (urlObj.changefreq) xmlLines.push(`    <changefreq>${urlObj.changefreq}</changefreq>`);
          if (urlObj.priority) xmlLines.push(`    <priority>${urlObj.priority}</priority>`);
          xmlLines.push('  </url>');
        });
      }
    } catch(e) {
      console.error('[Warning] guides_sitemap.json is invalid, skipping guides in sitemap.');
    }
  }

  xmlLines.push('</urlset>');

  fs.writeFileSync(SITEMAP_PATH, xmlLines.join('\n'), 'utf-8');
  console.log('[Success] sitemap.xml updated successfully!');
}


// syncNewsIndex()는 제거했다.
// index.html의 뉴스 카드는 auto_fetch_news.js가 news_list_*.json으로 4시간마다 갱신한다.
// 이 파일이 보던 data/news_list.json은 2026-07-16에서 멈춘 낡은 데이터여서,
// 사이트맵만 갱신하려고 이 스크립트를 돌리면 3개 언어 홈페이지 뉴스가
// 한 달 전 기사로 덮어써지는 사고가 났다. 이 스크립트는 사이트맵 생성 전용이다.

generateSitemap();
